pkg/utils/apply/apply.go

```diff
@@ -0,0 +1,205 @@
+package apply
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// An Object is a Kubernetes object.
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+// Applicator provides two approaches to applies changes to an object,
+// patching and updating.
+type Applicator interface {
+	PatchingApplicator
+	UpdatingApplicator
+}
+
+// PatchingApplicator applies changes to an object with patching.
+type PatchingApplicator interface {
+	PatchingApply(context.Context, runtime.Object, ...ApplyOption) error
+}
+
+// PatchingApplyFn implements PatchingApplicator.
+type PatchingApplyFn func(context.Context, runtime.Object, ...ApplyOption) error
+
+// PatchingApply applies changes to the supplied object.
+func (fn PatchingApplyFn) PatchingApply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
+	return fn(ctx, o, ao...)
+}
+
+// UpdatingApplicator applies changes to an object with updating.
+type UpdatingApplicator interface {
+	UpdatingApply(context.Context, runtime.Object, ...ApplyOption) error
+}
+
+// UpdatingApplyFn implements UpdatingApplicator.
+type UpdatingApplyFn func(context.Context, runtime.Object, ...ApplyOption) error
+
+// UpdatingApply applies changes to the supplied object.
+func (fn UpdatingApplyFn) UpdatingApply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
+	return fn(ctx, o, ao...)
+}
+
+// ApplyOption is called before applying the existing object to match the
+// desired object. ApplyOptions are still called even if the object does NOT exist.
+// If the object does not exist, `existing` will be assigned as `nil`.
+// nolint: golint
+type ApplyOption func(ctx context.Context, existing, desired runtime.Object) error
+
+// NewAPIApplicator creates returns an Applicator that applies changes to an
+// object by either creating or updating/patching it in a Kubernetes API server.
+func NewAPIApplicator(c client.Client) *APIApplicator {
+	return &APIApplicator{
+		creatorFn(createOrGetExisting),
+		patcherFn(threeWayMergePatch),
+		c}
+}
+
+type creator interface {
+	createOrGetExisting(context.Context, client.Client, runtime.Object, ...ApplyOption) (runtime.Object, error)
+}
+
+type creatorFn func(context.Context, client.Client, runtime.Object, ...ApplyOption) (runtime.Object, error)
+
+func (fn creatorFn) createOrGetExisting(ctx context.Context, c client.Client, o runtime.Object, ao ...ApplyOption) (runtime.Object, error) {
+	return fn(ctx, c, o, ao...)
+}
+
+type patcher interface {
+	calculatePatch(c, m runtime.Object) (client.Patch, error)
```

Comment:
the name should be `threeWayMergePatch`.  The `three way merge patch` is a famous name in K8s.

Reply:
got it.